src/visitors/derivative_original_visitor.cpp

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2023 Blue Brain Project, EPFL.
+ * See the top-level LICENSE file for details.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "visitors/derivative_original_visitor.hpp"
+
+#include "ast/all.hpp"
+#include "lexer/token_mapping.hpp"
+#include "pybind/pyembed.hpp"
+#include "utils/logger.hpp"
+#include "visitors/visitor_utils.hpp"
+#include <optional>
+#include <utility>
+
+namespace pywrap = nmodl::pybind_wrappers;
+
+namespace nmodl {
+namespace visitor {
+
+
+void CvodeVisitor::visit_derivative_block(ast::DerivativeBlock& node) {
+    node.visit_children(*this);
+    der_block = std::shared_ptr<ast::DerivativeBlock>(node.clone());
+}
+
+
+void CvodeVisitor::visit_cvode_block(ast::CvodeBlock& node) {
+    derivative_block = true;
+    node.visit_children(*this);
+    derivative_block = false;
+}
+
+void CvodeVisitor::visit_diff_eq_expression(ast::DiffEqExpression& node) {
+    differential_equation = true;
+    node.visit_children(*this);
+    differential_equation = false;
+}
+
+
+void CvodeVisitor::visit_binary_expression(ast::BinaryExpression& node) {
+    const auto& lhs = node.get_lhs();
+
+    /// we have to only solve ODEs under original derivative block where lhs is variable
+    if (!derivative_block || !differential_equation || !lhs->is_var_name()) {
+        return;
+    }
+
+    auto name = std::dynamic_pointer_cast<ast::VarName>(lhs)->get_name();
+
+    if (name->is_prime_name()) {
+        auto varname = "D" + name->get_node_name();
+        logger->debug("DerivativeOriginalVisitor :: replacing {} with {} on LHS of {}",
+                      name->get_node_name(),
+                      varname,
+                      to_nmodl(node));
+        node.set_lhs(std::make_shared<ast::Name>(new ast::String(varname)));
+        if (program_symtab->lookup(varname) == nullptr) {
+            auto symbol = std::make_shared<symtab::Symbol>(varname, ModToken());
+            symbol->set_original_name(name->get_node_name());
+            program_symtab->insert(symbol);
+        }
+    }
+}
+
+void CvodeVisitor::visit_program(ast::Program& node) {
+    program_symtab = node.get_symbol_table();
+    node.visit_children(*this);
+    if (der_block) {
+        auto der_node = new ast::CvodeBlock(der_block->get_name(),
+                                            der_block->get_statement_block(),
+                                            der_block->get_statement_block());
+        node.emplace_back_node(der_node);
+    }
```

Comment:
If you don't like the pattern that you need to toggle two booleans there's a way out. The code I marked essentially does a:

            auto blocks = collect_nodes(*ast, {ast::AstNodeType::DERIVATIVE_BLOCK});

Then all you need to do is run the `{}' -> d{}` replacement visitor (which consists of the `visit_binary_expression`) on each block and put them into `CVodeBlock`s and add them to the program.

Then there's no more boolean toggling; and it'll be more explicit about handling multiple DERIVATIVE blocks. BTW, the name for toggling being inside a cvode block is now rather misleading.